### ISSUE TITLE
Cabal scripts

### ISSUE THREAD

**Issue by georgefst** [haskell/haskell-language-server#111]
Certain upcoming features ([repl](https://github.com/haskell/cabal/issues/6149), [runghc](https://github.com/haskell/cabal/issues/6647)) are likely to make [cabal scripts](https://www.haskell.org/cabal/users-guide/nix-local-build.html#cabal-v2-run) a much more attractive option.

I, for one, would like to switch all of my `turtle` scripts to use them, rather than (as at present) having them all in a single cabal project. But I wouldn't want to lose the IDE support...

Would there be any major difficulty in supporting these? Is this actually more of a question for `hie-bios`?

**Comment by fendor** [comment]
Supporting cabal script is a matter of implementing the issue: https://github.com/haskell/cabal/issues/6149

That is why stack scripts are already supported via hie-bios.

I started to implement it, but there are a few roadbumps that makes it not as easy as hoped.

**Comment by jneira** [comment]
As commented in https://github.com/haskell/ghcide/issues/162 and here, there is a workaround to make stack scripts work:

Stack scripts work in a "hacky" way.
You have to add an explicit hie.yaml file with the contents:

```
cradle:
  stack:
    component: "File.hs" # your local stack script
```

Works fine for loading and modifying the stack script.
However, if you modify the script section, these changes wont be detected, e.g. you have to restart ghcide/hls.

The reason this works is that you can do stack repl File.hs and the session is loaded correctly.

by @fendor

**Comment by jneira** [comment]
Related issue in hi-bios: https://github.com/mpickering/hie-bios/issues/217

**Comment by fendor** [comment]
Afaik, `stack repl File.hs` ceased to work for stack scripts, therefore it does not work anymore with HLS :( 
I dont remember in which stack version it used to work (before 1.9.3?) and I cant find any related stack issues for it

**Comment by georgefst** [comment]
FWIW, I have been making a lot of use of Cabal scripts, using `cabal install --package-env . --lib [...]`. It's not ideal, but I can live with it for now. Doesn't even require a cradle. I don't think this worked when I opened the issue, but I can't remember why.

If it weren't for haskell/cabal#6999 things would actually be rather nice.

**Comment by jneira** [comment]
@Ailrun this one would be related with #692 (so it would be included for gsoc as well in the same pack)

**Comment by Ailrun** [comment]
Would it be better to create a new combined issue for GSOC?

**Comment by jneira** [comment]
Mmm not sure being only 2, it is clear in each one that are related

**Comment by rihardsk** [comment]
https://github.com/haskell/cabal/issues/6149 has just landed (there hasn't been a release yet, though). Does that mean support for Cabal scripts could be added now?

**Comment by fendor** [comment]
@rihardsk we have to look into it, but depending on the actual implementation, it could work now out of the box, assuming you have a `hie.yaml` with the contents:
```yaml
cradle:
  cabal:
```

right next to it.

**Comment by rihardsk** [comment]
@fendor it doesn't seem to work out of the box. I built cabal HEAD and prepended it to $PATH. I also created the `hie.yaml` as you suggested. When i open the cabal script with
```haskell
#! /usr/bin/env cabal        
{- cabal:                    
build-depends:               
    base ^>= 4.14            
  , random-fu ^>= 0.2.7.7    
  , mwc-random ^>= 0.15.0.2  
  , split ^>= 0.2.3.4        
  , pipes ^>= 4.3.16         
-}                           

-- ...
```
in the header, hls fails to start and i get the following output in the `*lsp-haskell::stderr*` buffer in `emacs`:
```
Found "/home/rihards/darbs/exp/adaptive-marian/code/hie.yaml" for "/home/rihards/darbs/exp/adaptive-marian/code/a"                                                                                                                             
Run entered for haskell-language-server-wrapper(haskell-language-server-wrapper) Version 1.5.1.0, Git revision 745ef26f406dbdd5e4a538585f8519af9f1ccb09 (dirty) x86_64 ghc-8.10.7                                                              
Current directory: /home/rihards/darbs/exp/adaptive-marian/code                                                                                                                                                                                
Operating system: linux                                                                                                                                                                                                                        
Arguments: ["--lsp","-d","-l","/tmp/hls.log"]                                                                                                                                                                                                  
Cradle directory: /home/rihards/darbs/exp/adaptive-marian/code                                                                                                                                                                                 
Cradle type: Cabal                                                                                                                                                                                                                             
                                                                                                                                                                                                                                               
Tool versions found on the $PATH                                                                                                                                                                                                               
cabal:    3.7                                                                                                                                                                                                                                  
stack:    2.7.3                                                                                                                                                                                                                                
ghc:    8.10.7                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                               
Consulting the cradle to get project GHC version...                                                                                                                                                                                            
Failed to get project GHC version:CradleError {cradleErrorDependencies = [], cradleErrorExitCode = ExitFailure 1, cradleErrorStderr = ["Error when calling cabal --builddir=/home/rihards/.cache/hie-bios/dist-code-b81b1ed796ad78aa090f9ea451b44b34 v2-exec --with-compiler /home/rihards/.cache/hie-bios/wrapper-13a09b18ea883dd377d59db5e821a86b ghc -v0 -- --numeric-version","","No cabal.project file or cabal file matching the default glob './*.cabal' was found.\nPlease create a package description file <pkgname>.cabal or a cabal.project file referencing the packages you want to build.\n\n"]}                                                                                                                            
                                                                                                                                                                                                                                               
Process lsp-haskell stderr finished                                                                                                                                                                                                            
```

`/tmp/hls.log` doesn't get any new entries as a result of the above, btw.

I checked that the new `cabal repl add-tags.hs` command does indeed work and brings up a `ghci` session with my stuff.

**Comment by fendor** [comment]
Thank you for the report!

Sounds like `cabal v2-exec ghc -v0 -- --numeric-version` basically fails to run.
We need to revive https://github.com/haskell/cabal/pull/7500 and push it over the finish line >_>

**Comment by dschrempf** [comment]
> @rihardsk we have to look into it, but depending on the actual implementation, it could work now out of the box, assuming you have a `hie.yaml` with the contents:
> 
> ```yaml
> cradle:
>   cabal:
> ```
> 
> right next to it.
> 
> You can help checking it out, by installing cabal HEAD and adding the aforementioned `hie.yaml`. If it works, hooray, otherwise feel free to provide log messages :)

Thank you!

This actually worked for me wit `hls-1.8` and `ghc924`. It would be great though, if the somewhat useless `hie.yaml` file could be omitted!

**Comment by fendor** [comment]
@dschrempf At the current time, the `hie.yaml` is the way to tell HLS how the project shall be loaded. If there is no cabal.project, no *.cabal file, no stack.yaml, how is it supposed to guess you want to use cabal on it?

**Comment by dschrempf** [comment]
I understand that hls needs this information. But the file could be a fallback or so?

**Comment by fendor** [comment]
A fallback for what case? It is optional, but for cabal scripts, HLS currently has no way of determining that the given file is a cabal script (we could parse the bang line I suppose) to guess correctly what to use to load the script.

**Comment by dschrempf** [comment]
Well, I do not know how HLS determines the file type, libraries etc. But from an outsider's point of view it seems pretty stupid that I have to provide a file that has essentially no information in it to make things work :smile:. The only point in having the `hie.yaml` file I could see is that the file deviates the attention of HLS from another project up on the directory structure to the current directory. Is that so?

**Comment by fendor** [comment]
> has essentially no information in it to make things work

It does have information, namely that it is a file that can be loaded via cabal. You could also have a stack project, a stack script, or a plain Haskell script as well. It is not trivial to know and `hie.yaml` disambiguates for that case.

> Is that so?

No, `hie.yaml` basically tells HLS exactly how to build a project, the reason it is so small is that we can infer the rest from the context once we know we should use cabal. In fact, every time HLS opens a project, it generates a hie.yaml internally (a cradle, but hie.yaml is a cradle description) and uses it to load the project. Again, maybe we could improve the discovery mechanism, but there are many cases where an explicit hie.yaml is just necessary.

**Comment by dschrempf** [comment]
I see. In thin case I advocate for improving the discovery mechanism. Thanks for explaining things to me!

**Comment by georgefst** [comment]
> This actually worked for me wit `hls-1.8` and `ghc924`

Not working for me unfortunately, with the same tools (Cabal 3.8.1.0, Linux).

> It would be great though, if the somewhat useless `hie.yaml` file could be omitted!

I think "useless" is an exaggeration, as @fendor has said, but I agree that we should strive to make it unnecessary. I was also going to suggest parsing the bang line. There are potentially a lot of edge cases, but recognising `#!/usr/bin/env cabal` would be a start.

**Comment by dschrempf** [comment]
> Not working for me unfortunately, with the same tools (Cabal 3.8.1.0, Linux).

It could be that this works for me because the script is within a project that provides the required libraries.

**Comment by georgefst** [comment]
Just to confirm, no one's actually got this to work properly by specifying a `hie.yaml`, right? It seems unlikely that it's possible, right now, since HLS runs `cabal exec -v0 -- ghc --print-libdir`, which complains about not finding a `*.cabal` or `cabal.project` file.

So does that mean we're still blocked on haskell/cabal#7500? Or is there, in principal, some easy way to extract the environment information from `cabal repl`?

**Comment by fendor** [comment]
The former can be fixed via the linked PR.
I don't think there is currently a way to get the options from `cabal repl`. However, we could tweak `hie-bios` to have a cli interface to get the options if and only if you specify the ghc version you want to use for scripts. Such a cli interface, could also be used to implement proper support... PRs welcome, won't be able to get to it the next couple of weeks, but I think the POC might be much simpler now.

**Comment by georgefst** [comment]
I think the reason for the confusion above is that, in addition to the `hie.yaml`, a `cabal.project` file is currently required (whose contents don't seem to matter much). @fendor pointed this out in https://github.com/haskell/haskell-language-server/discussions/3572.

So loading scripts does now work, though the workarounds required are quite inconvenient if one has a lot of scripts.

**Comment by georgefst** [comment]
> a `cabal.project` file is currently required

I don't know when this changed, but this is seemingly no longer needed! The boilerplate `hie.yaml` still is though.

**Comment by subterfugue** [comment]
@georgefst Do you have any idea about how that came to be? It looks like two alternatives to that PR that was mentioned might elide the need of a `cabal.project` file: one for `cabal v2-path`, and another for `cabal target`. The latter requires a `cabal.project`, but the former doesn't. Could that be what HLS uses now to determine the flags, or is it something else?

Putting that aside, is modifying `hie-bios` to have a CLI interface to get the required flags still the main way to go about it? It was stated earlier in the discussion that HLS doesn't know how to differentiate between stack scripts and normal files, but `cabal run` has the ability to do that, for example, as does `stack`. Is there no way for them to expose the methods they use to check and use it in HLS, or just shell out to them instead if that's not what's already being done? Sorry for my ignorance, I'm not sure about how any of these projects work.

**Comment by georgefst** [comment]
No idea, sorry. I don't think I've been personally involved in the implementation of any of this.

**Comment by fendor** [comment]
> Could that be what HLS uses now to determine the flags, or is it something else?

It could be!

> Putting that aside, is modifying hie-bios to have a CLI interface to get the required flags still the main way to go about it?

I don't think so, doesn't it almost work now out of the box, the only thing required being a 

```yaml
cradle:
  cabal:
```

We could try to refine the session guessing in `HLS`, if the module has a cabal script marker such as 

```
#!/usr/bin/env cabal
{- cabal:
build-depends:
  base,
  haskell-say ^>=1.0.0.0
-}
```

then we infer the correct cradle. Such changes would be in `ghcide/session-loader/Development/IDE/Session/Implicit.hs`, perhaps it is easy enough to do now?

How does cabal check for these script markers in Haskell files?

**Comment by upsole** [comment]
Looks like cabal runs a simple [parsing](https://github.com/haskell/cabal/blob/master/cabal-install/src/Distribution/Client/ScriptUtils.hs#L552) function to figure out the script metadata.

We could use something similar to build a predicate and scan `start_dir` for a file that matches as the lowest priority option in this [cradle guesser](https://github.com/haskell/haskell-language-server/blob/master/ghcide/session-loader/Development/IDE/Session/Implicit.hs#L58). `simpleCabalCradle` seems to do the trick.

Would love to avoid the scan but it seems like `loadImplicitCradle` only ever gets called with a directory.

What do you think?

**Comment by wz1000** [comment]
I suspect if you could get `cabal repl myscript.hs` to work, then HLS would "just work" with the file as well.

**Comment by fendor** [comment]
It would still require an explicit `hie.yaml` to tell hie-bios that it should interpreter it as a cabal cradle, so I am wondering whether we should add some limited parsing to hie-bios, e.g. bang pattern parsing to decide which cradle to use.

**Comment by upsole** [comment]
As far as I can tell, the logic deciding what happens in the absence of `hie.yaml` lives in `ghcide/session-loader/Development/IDE/Session/Implicit.hs`, so that's been my first guess as to where the parsing should happen. Do you mean you'd rather have  this handled completely by `hie-bios` ?

I've ran locally with this and it results on `hie-bios` compiling the script with `v2-repl`.

``` haskell
inferCradleTree :: FilePath -> MaybeT IO (CradleTree a, FilePath)
inferCradleTree start_dir =
       maybeItsBios
   -- If we have both a config file (cabal.project/stack.yaml) and a work dir
   -- (dist-newstyle/.stack-work), prefer that
   <|> (cabalExecutable >> cabalConfigDir start_dir >>= \dir -> cabalWorkDir dir >> pure (simpleCabalCradle dir))
   <|> (stackExecutable >> stackConfigDir start_dir >>= \dir -> stackWorkDir dir >> stackCradle dir)
   -- If we have a cabal.project OR we have a .cabal and dist-newstyle, prefer cabal
   <|> (cabalExecutable >> (cabalConfigDir start_dir <|> cabalFileAndWorkDir) <&> simpleCabalCradle)
   -- If we have a stack.yaml, use stack
   <|> (stackExecutable >> stackConfigDir start_dir >>= stackCradle)
   -- If we have a cabal file, use cabal
   <|> (cabalExecutable >> cabalFileDir start_dir <&> simpleCabalCradle)
   -- If we have a haskell cabal script, use cabal
   <|> (maybeHoldsCabalScript start_dir <&> simpleCabalCradle)
```

Stack is another story as the currently existing functions seem to assume a `stack.yaml` being present.

It's possibly too blunt of a solution as I can imagine I'm missing a lot of context

**Comment by fendor** [comment]
Yes, I think that is a viable solution. I would like to see this check in hie-bios as well, since I often use hie-bios to debug cradles.

So, the shebang parser should be in hie-bios and HLS should use it here as well.

`stack script`s don't need a `stack.yaml`, iirc, this would be completely counterproductive. https://docs.haskellstack.org/en/stable/commands/script_command/

In theory (and practice), scripts could be polyglots for cabal and stack, e.g. this:
```haskell
#! /usr/bin/env stack
{- stack script
   --snapshot lts-24.43
   --package haskell-say
-}
{- cabal:
build-depends:
  base,
  haskell-say ^>=1.0.0.0
-}

import HaskellSay

main :: IO ()
main = haskellSay "Test"

```
can be run with both `stack Main.hs` and `cabal run Main.hs` (or `cabal repl Main.hs`).

Question is, do we want to support this, what is the cabal spec for when the magic executable stanza  is accepted in the comment, or should we rely purely on the shebang line?

It does look like both tools only need the magic comment block, so perhaps our inference could be just reading the shebang line to decide whether this is a cabal or stack script.
